### PR TITLE
Add documentation for CrossGen

### PR DESCRIPTION
**Pull request by JohnChen0**

### CODE REVIEW

**Comment by JohnChen0** [comment]
@gkhanna79: PTAL

File: Documentation/building/crossgen.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+- If you install crossgen from NuGet, make sure you use crossgen from exactly the same NuGet package as the rest of you CoreCLR binaries. Do not attempt to mix binaries from multiple NuGet packages.
+
+If you do not follow the above rules, you are likely to encounter errors while running crossgen.
+
+Using crossgen
+--------------
+
+In most cases, the build script automatically runs crossgen to create the native image for mscorlib.
+When this happens, you will find boths `mscorlib.dll` and `mscorlib.ni.dll` in your output directory.
+`mscorlib.dll` is the MSIL assembly creating by the C# compiler, while `mscorlib.ni.dll` is the native image that contains CPU-specific code.
```
Comment (alaatm) [inline Documentation/building/crossgen.md]: Typo. 
"...is the MSIL assembly creating ..." 
should be 
"...is the MSIL assembly created ..."

File: Documentation/building/crossgen.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
+Using CrossGen to Create Native Images
+======================================
+
+Introduction
+------------
+
+When you create a .NET assembly using C# compiler, your assembly contains only MSIL code.
+When the app runs, the JIT compiler translates the MSIL code into native code, before the CPU can execute them.
+This execution model has some advantages. For example, your assembly code can be portable across all platforms and architectures that supports .NET Core.
```
Comment (ellismg) [inline Documentation/building/crossgen.md]: "supports" -> "support"

File: Documentation/building/crossgen.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+However, this portability comes with a performance penalty. Your app starts up slower, because the JIT compiler has to spend time to translate the code.
+
+To help make your app start up faster, CoreCLR includes a tool called crossgen, which can pre-compile the MSIL code into native code.
+
+Getting crossgen
```
Comment (bendono) [inline Documentation/building/crossgen.md]: Other than the filename, casing should be consistent. As used above, assuming "CrossGen" is the preferred spelling.

File: Documentation/building/crossgen.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+----------------
+
+If you build CoreCLR yourself, the crossgen tool (`crossgen.exe` on Windows, or `crossgen` on other platforms) is created as part of the build, and stored in the same output directory as other CoreCLR binaries.
```
Comment (bendono) [inline Documentation/building/crossgen.md]: ... the CrossGen tool...

File: Documentation/building/crossgen.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+If you install CoreCLR using a NuGet package, you can find crossgen in the `tools` folder of the NuGet package.
```
Comment (bendono) [inline Documentation/building/crossgen.md]: ... you can find CrossGen...

File: Documentation/building/crossgen.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+Regardless of how you obtain crossgen, it is very important that it must match other CoreCLR binaries.
```
Comment (bendono) [inline Documentation/building/crossgen.md]: Regardless of how you obtain CrossGen...

File: Documentation/building/crossgen.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+- If you build crossgen yourself, you should use it with coreclr and mscorlib generated from the same build. Do not attempt to mix crossgen from one build with binaries generated from another build.
```
Comment (bendono) [inline Documentation/building/crossgen.md]: If you build CrossGen yourself...
Do not attempt to mix CrossGen...

File: Documentation/building/crossgen.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+- If you install crossgen from NuGet, make sure you use crossgen from exactly the same NuGet package as the rest of you CoreCLR binaries. Do not attempt to mix binaries from multiple NuGet packages.
```
Comment (bendono) [inline Documentation/building/crossgen.md]: If you install CrossGen...

File: Documentation/building/crossgen.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+If you do not follow the above rules, you are likely to encounter errors while running crossgen.
```
Comment (bendono) [inline Documentation/building/crossgen.md]: ... while running CrossGen.

File: Documentation/building/crossgen.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+Using crossgen
```
Comment (bendono) [inline Documentation/building/crossgen.md]: Using CrossGen

File: Documentation/building/crossgen.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+--------------
+
+In most cases, the build script automatically runs crossgen to create the native image for mscorlib.
```
Comment (bendono) [inline Documentation/building/crossgen.md]: ... automatically runs CrossGen...

File: Documentation/building/crossgen.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+When this happens, you will find boths `mscorlib.dll` and `mscorlib.ni.dll` in your output directory.
+`mscorlib.dll` is the MSIL assembly creating by the C# compiler, while `mscorlib.ni.dll` is the native image that contains CPU-specific code.
+Once the build is done, you only need `mscorlib.ni.dll` to use CoreCLR.
+As a matter of fact, most CoreCLR NuGet packages contain only `mscorlib.ni.dll`, without `mscorlib.dll`
+
+If for some reason you did not get `mscorlib.ni.dll` with the rest of your CoreCLR, you can easily create it yourself using crossgen.
```
Comment (bendono) [inline Documentation/building/crossgen.md]: ... using CrossGen.

File: Documentation/building/crossgen.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    ./crossgen /Platform_Assemblies_Paths "path1;path2;..." assemblyName.dll
+    
+The /Platform_Assemblies_Paths is used to specify the location of all the dependencies of the input assembly.
+You should use full paths for this locations. Relative paths do not always work.
+If there are multiple paths, separate them with semicolons (`;`). On non-Windows platforms, be sure to enclose the path list in quotes to protect the semicolons from the shell.
+
+Using native images
+-------------------
+
+Running crossgen on an assembly creates a "native image" file, with the extension of `.ni.dll` or `.ni.exe`.
```
Comment (bendono) [inline Documentation/building/crossgen.md]: Running CrossGen on...

File: Documentation/building/crossgen.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+You should include the native images in your app, at the same location where you normally install the MSIL assemblies.
+Once you have included native images, you do not need to include the original MSIL assemblies in your apps.
+
+Common errors
+-------------
+
+The following are some of the command errors while creating or using native images:
+- "Could not load file or assembly 'mscorlib.dll' or one of its dependencies. The native image could not be loaded, because it was generated for use by a different version of the runtime. (Exception from HRESULT: 0x80131059)": This error indicates that there is a mismatch between crossgen and mscorlib.ni.dll. Make sure use crossgen and mscorlib.ni.dll from the same build or NuGet package.
```
Comment (bendono) [inline Documentation/building/crossgen.md]: ... there is a mismatch between CrossGen and...
Make sure **to** use CrossGen...

**Comment by JohnChen0** [comment]
Thanks for the comments. The PR has been updated accordingly.

**Pull request merged**